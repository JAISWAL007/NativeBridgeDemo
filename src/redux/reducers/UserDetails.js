import {USER_DETAILS} from '../actionTypes';

const initialState = {
  username: 'sdc',
};

export default function (state = initialState, action) {
  switch (action.type) {
    case USER_DETAILS: {
      return {
        ...state,
        name: action.payload,
      };
    }
    default:
      return state;
  }
}
