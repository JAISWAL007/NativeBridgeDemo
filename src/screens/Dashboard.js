import React from 'react';
import {
  SafeAreaView,
  StyleSheet,
  Text,
  NativeModules,
  View,
  TouchableOpacity,
} from 'react-native';
import PropTypes from 'prop-types';
import {useSelector} from 'react-redux';
import Icon from 'react-native-vector-icons/FontAwesome';

import Strings from '../Constant';

const DummyScreen = ({navigation}) => {
  const [isEmulator, setIsEmulator] = React.useState('');
  const storeUserName = useSelector((state) => state?.UserDetails?.name);

  const checkIsEmulator = async () => {
    const {DeviceDetails} = NativeModules;
    const emulator = await DeviceDetails.isEmulator();
    setIsEmulator(emulator);
  };

  React.useEffect(() => {
    checkIsEmulator();
  }, []);
  return (
    <SafeAreaView style={styles.container}>
      <View style={styles.subContainer}>
        <TouchableOpacity
          onPress={() => navigation.goBack()}
          testID="back-button">
          <Icon name="angle-left" color="#000" size={30} />
        </TouchableOpacity>
        <Text style={styles.title}>{Strings.DASHBOARD}</Text>
        <TouchableOpacity
          testID="next-button"
          onPress={() => navigation.navigate('ButtonScreen')}>
          <Icon name="angle-right" color="#000" size={30} />
        </TouchableOpacity>
      </View>
      <View style={styles.displayStyle}>
        <Text>Welcome {storeUserName}</Text>
        <Text testID="displayLabel">
          Running on {isEmulator ? 'Emulator' : 'Devices'}
        </Text>
      </View>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  displayStyle: {
    marginTop: 100,
    justifyContent: 'center',
    alignItems: 'center',
  },
  item: {
    backgroundColor: '#fff',
    padding: 20,
    marginVertical: 8,
    marginHorizontal: 20,
  },
  title: {
    fontSize: 32,
    textAlign: 'center',
  },
  subContainer: {
    flexDirection: 'row',
    justifyContent: 'space-around',
    alignItems: 'center',
  },
  subTitle: {
    fontSize: 20,
    marginTop: 10,
  },
});

DummyScreen.prototype = {
  navigation: PropTypes.objectOf({
    navigate: PropTypes.func,
    goBack: PropTypes.func,
  }).isRequired,
};
export default DummyScreen;
