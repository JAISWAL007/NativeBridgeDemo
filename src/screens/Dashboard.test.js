import React from 'react';
import {NativeModules} from 'react-native';
import {
  fireEvent,
  flushMicrotasksQueue,
  render,
} from '@testing-library/react-native';
import {Provider} from 'react-redux';
import {createStore} from 'redux';
import Dashboard from './Dashboard';
import reducers from '../redux/reducers';

describe('Dashboard Screen', () => {
  const mockNavigation = {
    navigate: jest.fn(),
    goBack: jest.fn(),
  };

  beforeEach(() => {
    const isEmulator = jest.fn(() => Promise.resolve(true));
    NativeModules.DeviceDetails = {
      isEmulator,
    };
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  const renderWithRedux = (
    component,
    {initialState, store = createStore(reducers, initialState)} = {},
  ) => {
    return {
      ...render(<Provider store={store}>{component}</Provider>),
    };
  };

  it('Should render Dashboard screen', async () => {
    const tree = renderWithRedux(<Dashboard navigation={mockNavigation} />);
    await flushMicrotasksQueue();
    expect(tree).toMatchSnapshot();
  });

  it('Should able to go back when user click on back arrow button', async () => {
    const {getByTestId} = renderWithRedux(
      <Dashboard navigation={mockNavigation} />,
    );
    await flushMicrotasksQueue();
    const backButton = getByTestId('back-button');
    await fireEvent.press(backButton);
    expect(mockNavigation.goBack).toBeCalledTimes(1);
  });

  it('Should able to navigate to button screen when user click on next button', async () => {
    const {getByTestId} = renderWithRedux(
      <Dashboard navigation={mockNavigation} />,
    );
    await flushMicrotasksQueue();
    const homeButton = getByTestId('next-button');
    await fireEvent.press(homeButton);
    expect(mockNavigation.navigate).toBeCalledTimes(1);
  });

  it('Should able to call native bridge and display when emulator', async () => {
    const {getByTestId} = renderWithRedux(
      <Dashboard navigation={mockNavigation} />,
    );
    await flushMicrotasksQueue();
    const label = getByTestId('display-label');
    expect(label.props.children).toStrictEqual(['Running on ', 'Emulator']);
  });

  it('Should able to call native bridge and display when Device', async () => {
    jest.clearAllMocks();
    const isEmulator = jest.fn(() => Promise.resolve(false));
    NativeModules.DeviceDetails = {
      isEmulator,
    };
    const {getByTestId} = renderWithRedux(
      <Dashboard navigation={mockNavigation} />,
    );
    await flushMicrotasksQueue();
    const label = getByTestId('display-label');
    expect(label.props.children).toStrictEqual(['Running on ', 'Devices']);
  });

  it('Should render with redux with custom store', async () => {
    const {getByTestId} = renderWithRedux(
      <Dashboard navigation={mockNavigation} />,
    );
    await flushMicrotasksQueue();
    const label = getByTestId('store-label');
    expect(label.props.children).toStrictEqual(['Welcome ', 'dummy name']);
  });
});
