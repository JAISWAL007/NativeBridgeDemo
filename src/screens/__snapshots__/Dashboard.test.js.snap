// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard Screen Should render Dashboard screen 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-around",
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="back-button"
    >
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "color": "#000",
              "fontSize": 30,
            },
            undefined,
            Object {
              "fontFamily": "FontAwesome",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        
      </Text>
    </View>
    <Text
      style={
        Object {
          "fontSize": 32,
          "textAlign": "center",
        }
      }
    >
      Dashboard
    </Text>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="next-button"
    >
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "color": "#000",
              "fontSize": 30,
            },
            undefined,
            Object {
              "fontFamily": "FontAwesome",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "marginTop": 100,
      }
    }
  >
    <Text
      testID="store-label"
    >
      Welcome 
      dummy name
    </Text>
    <Text
      testID="display-label"
    >
      Running on 
      Emulator
    </Text>
  </View>
</RCTSafeAreaView>
`;
